File: arrays-and-strings/122. Best Time to Buy and Sell Stocks II/solution.ts
function maxProfit(prices: number[]): number {
  let minPrice = Infinity;
  let cumulativeProfit = 0;
  let maxCurrentProfit = 0;
  let currentProfit = 0;

  for (const p of prices) {
    if (p < minPrice) minPrice = p;
    currentProfit = p - minPrice;
    if (currentProfit > 0) {
      cumulativeProfit += currentProfit;
      minPrice = p;
    }
  }
  return cumulativeProfit;
}

export { maxProfit };

File: arrays-and-strings/122. Best Time to Buy and Sell Stocks II/solution.py
class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        if(len(prices) < 2):
            return 0

        profit = 0
        best_buy = best_sell = prices[0]

        for p in prices:
            if (p < best_buy):
                best_buy = p
                best_sell = p
            elif(p > best_sell):
                profit += p - best_buy
                best_buy = p
                best_sell = 0
        return profit